File: app/src/main/java/io/weimu/www/imagepicker/activity/MainActivity.java
package io.weimu.www.imagepicker.activity;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.tbruyelle.rxpermissions2.RxPermissions;
import com.weimu.library.ImagePicker;
import com.weimu.library.utils.GridSpacingItemDecoration;
import com.weimu.library.utils.ScreenUtils;
import com.weimu.library.view.ImageSelectorActivity;

import java.util.ArrayList;

import io.reactivex.functions.Consumer;
import io.weimu.www.imagepicker.R;
import io.weimu.www.imagepicker.adaper.recycleview.ImageGridadapter;

public class MainActivity extends AppCompatActivity {


    private RecyclerView recyclerView;
    private ImageGridadapter mAdapter;
    private GridLayoutManager gridManager;


    private int maxImageNumber = 9;
    private int spanCount = 4;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        new RxPermissions(this)
                .request(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.CAMERA)
                .subscribe(new Consumer<Boolean>() {
                    @Override
                    public void accept(Boolean granted) throws Exception {
                        if (granted)
                            initRecyclerVIew();

                    }
                });


    }

    private void initRecyclerVIew() {
        recyclerView = (RecyclerView) findViewById(R.id.id_RecyclerView);
        mAdapter = new ImageGridadapter(this, maxImageNumber);
        gridManager = new GridLayoutManager(this, spanCount);
        recyclerView.setLayoutManager(gridManager);
        recyclerView.addItemDecoration(new GridSpacingItemDecoration(spanCount, ScreenUtils.dip2px(this, 8), false));
        //设置Item增加、移除动画
        //recyclerView.setItemAnimator(new NoAlphaItemAnimator());
        recyclerView.setAdapter(mAdapter);

        mAdapter.AddListenter(new ImageGridadapter.AddListenter() {
            @Override
            public void onAddClick(int needNumber) {
                //ImagePicker.getInstance().pickAvatar(MainActivity.this);
                ImagePicker.getInstance().pickImage(MainActivity.this, 9);
                //ImagePicker.getInstance().takePhoto(MainActivity.this,true);//使用摄像头
            }

            @Override
            public void onItemClick(int position) {
                startActivity(PhotoViewPagerActivity.newInstance(MainActivity.this, position, (ArrayList<String>) mAdapter.getDataList()));
            }

            @Override
            public void onItemDeleteClick(int position) {
                mAdapter.deleteData(position);
            }
        });
    }


    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == ImageSelectorActivity.REQUEST_IMAGE) {
            ArrayList<String> pics = (ArrayList<String>) data.getSerializableExtra(ImageSelectorActivity.REQUEST_OUTPUT);
            mAdapter.addData(pics);
        }
    }
}

File: app/src/main/java/io/weimu/www/imagepicker/activity/MainActivity.kt
package io.weimu.www.imagepicker.activity

import android.Manifest
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import com.weimu.library.ImagePicker
import com.weimu.library.view.ImageSelectorActivity
import com.weimu.universalib.ktx.dip2px
import com.weimu.universalib.ktx.getColorPro
import com.weimu.universalview.core.recyclerview.decoration.GridItemDecoration
import com.weimu.universalview.core.toolbar.StatusBarManager
import com.weimu.universalview.ktx.requestPermission
import io.weimu.www.imagepicker.R
import io.weimu.www.imagepicker.fragment.adapter.ImageGridAdapter
import java.util.*

class MainActivity : AppCompatActivity() {


    private var recyclerView: RecyclerView? = null
    private var mAdapter: ImageGridAdapter? = null
    private var gridManager: GridLayoutManager? = null


    private val maxImageNumber = 9
    private val spanCount = 4

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        StatusBarManager.setColor(window, getColorPro(R.color.white))
        StatusBarManager.setLightMode(window)

        requestPermission(
                permissions = *arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE, Manifest.permission.CAMERA),
                granted = { initRecyclerVIew() },
                dialogMessage = "请给我权限！"
        )
    }

    private fun initRecyclerVIew() {
        recyclerView = findViewById(R.id.id_RecyclerView)
        mAdapter = ImageGridAdapter(this, maxImageNumber)
        gridManager = GridLayoutManager(this, spanCount)
        recyclerView!!.layoutManager = gridManager
        recyclerView?.addItemDecoration(GridItemDecoration(spanCount, dip2px(8f), dip2px(8f)))
        //设置Item增加、移除动画
        //recyclerView.setItemAnimator(new NoAlphaItemAnimator());
        recyclerView!!.adapter = mAdapter


        mAdapter?.imageActionListener = object : ImageGridAdapter.ImageActionListener {

            override fun onItemClick(position: Int) {
                startActivity(PhotoViewPagerActivity.newInstance(this@MainActivity, position, mAdapter!!.dataList as ArrayList<String>))
            }

            override fun onItemDeleteClick(position: Int) {
                mAdapter!!.removeItem(position)
            }
        }
        mAdapter?.onFooterClick={
            //                ImagePicker.pickAvatar(this@MainActivity);
            ImagePicker.pickImage(this@MainActivity, 9)
//                ImagePicker.takePhoto(this@MainActivity,true);//使用摄像头
        }
    }


    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == ImageSelectorActivity.REQUEST_IMAGE) {
            val pics = data!!.getSerializableExtra(ImageSelectorActivity.REQUEST_OUTPUT) as ArrayList<String>
            mAdapter!!.addData(pics)
        }
    }
}
